### PR TITLE
Expose utilization along channels in mz_message_counts

### LINKED ISSUE #4584: Surface logged timely dataflow message counts
Timely dataflow logs numbers of messages that move along each of the channels. This is information, not unlike memory and CPU used, that can help folks diagnose performance issues.

One particular thing I would like to see are the numbers of data vs progress messages. The latter has the potential to be quadratic in the worker count, and while there are interventions we can do we would want to see the measurements first.

### PR DESCRIPTION
**Pull request by antifuchs**
This PR fixes #4584 by adding a table `mz_message_counts` that contains information about how many messages have passed down a channel.

The structure of the table is now as follows:

```
 channel | source | target | sent  | received 
---------+--------+--------+-------+----------
```

Where `source` and `target` are the source&target workers, and the `sent` and `received` counts are the numbers of messages on the channel - if "sent" is way ahead of "received", that would probably indicate catastrophic behindness in the whole system?

cc @frankmcsherry & @ruchirK (:

---
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/5939)at

### CODE REVIEW

**Comment by frankmcsherry** [comment]
> There's a sent flag, so if we group by that, does that mean we could expose the number of sent/received messages in a channel, even? (Thus, see where operators fall behind on available data, maybe?)

You totally could! 

Some further thoughts: if you perform this subtraction in dataflow, it will have the property that the maintained state will vanish when there are no outstanding records (and generally be proportional to the number of channels on which there are outstanding records). But, also doing it in dataflow risks going negative (as reports from different workers are unsynchronized) and MZ falls over if you do that. So, maybe best to just group by that too.

Any interest in putting in the source and target worker ids as well?

> I'm running into a weird type error running this:

It looks like you are colliding with the identifier `3027`. It does seem awkward that one has to manually specify all of these; I think @justinj is working on efforts to robustify how the catalog names things, but idk if it is pointed at the problem of manually assigning identifiers.
```
pub const MZ_SOURCE_INFO: BuiltinLog = BuiltinLog {
    name: "mz_source_info",
    schema: MZ_CATALOG_SCHEMA,
    variant: LogVariant::Materialized(MaterializedLog::SourceInfo),
    id: GlobalId::System(3026),
    index_id: GlobalId::System(3027),
};
```

**Comment by ruchirK** [comment]
this looks great! i think adding the source + destination worker could be really useful too (I think this logging will maybe be most useful to find issues in exchange edges and being able to point to a particular worker or pair of workers could be really nice)

> But, also doing it in dataflow risks going negative (as reports from different workers are unsynchronized) and MZ falls over if you do that.

I'm not sure what would cause MZ to fall over if the (channel, current_count) in this log went negative could you say more about why that would happen? (I don't see any assertions for current_count >= 0, not sure if theres something else to look for)

**Comment by ruchirK** [comment]
oh whoops frank already said the thing about source and target workers lol

**Comment by frankmcsherry** [comment]
> I'm not sure what would cause MZ to fall over if the (channel, current_count) in this log went negative could you say more about why that would happen? (I don't see any assertions for current_count >= 0, not sure if theres something else to look for)

We used to have them in `reduce.rs`, but it looks like now we just log an error and produce potentially incorrect output. It used to panic in the collator stage of reduce, because one aggregate would be missing. I can't easily see what happens in the code today though.

**Comment by ruchirK** [comment]
oh right. I was imagining that we would do it in this pr directly, instead of relying on a group by statement. in my head it would have been something like

```
let messages = messages
            .as_collection()
            .explode(|(key, count, sent)| {
                if sent 
                   Some((key, count * -1 as i64))
                } else {
                   Some((key, count as i64))
                }
           })
           ....
```

but thinking about it more I'm not sure if that actually makes sense

**Comment by frankmcsherry** [comment]
Ah, I meant that if we publish a table with negative counts, that could cause MZ to be angry. If we take your fragment above and put a `.count()` at the end to ensure that we correctly handle the negativity ourselves, could be fine!

**Comment by antifuchs** [comment]
The way we ended up doing it is by using a `DiffPair`, which happens to match exactly the duality of thing that we want to look at (:

File: src/coord/tests/builtins.rs [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+            }
+            View(BuiltinView { name, .. }) => {
+                encounter("view", "id", name, id);
+            }
+            Func(BuiltinFunc { name, .. }) => {
+                encounter("function", "id", name, id);
+            }
+        }
+    }
+}
```
Comment (benesch) [inline src/coord/tests/builtins.rs:46]: Thanks very much for doing this! I have a suggestion, though! Can we perform the check inline when we construct the `BUILTINS` variable? As in, somewhere here: https://github.com/MaterializeInc/materialize/blob/476f8b593e8640870101528cac7766880b3deb7d/src/coord/src/catalog/builtin.rs#L1336

The idea being that this test will catch the duplicate ID bug if someone is running `cargo test`, but if they're testing their change via `cargo run`, I think they'd still be doomed to bad error messages?

Comment (antifuchs) [reply src/coord/tests/builtins.rs:46]: Yep, that makes sense - I guess it will cause tests to fail too, incidentally (not sure how lazy_static will do its thing), so if things fall the way I imagine they will, I'll leave a test (but not this test) in there to ensure it triggers in CI, too.

Comment (antifuchs) [reply src/coord/tests/builtins.rs:46]: (Now that I think about it, there's probably no realistic branch of the universe on which our CI doesn't set up the builtin lazy_static initialization. I'll drop the test) (:

Comment (benesch) [reply src/coord/tests/builtins.rs]: Yup exactly. I can imagine other situations where you'd want to write a test like

```
#[test]
fn test_touch_static() {
  black_box(&*BUILTINS)
}
```

but here I agree that it's quite safe to rely on the many tests that actually start Materialize to run the BUILTINS initializer.

**Comment by antifuchs** [comment]
Soo, hm, @frankmcsherry - I had to do some surgery on the `channels` collection structure so that joining with it wouldn't turn ridiculous (unpacking / typecasting out of a Row into a usize again). But unfortunately, it looks like with the latest change in 540b66f4dcad16c72a858ff07b6c64a5897e31b4, it's not thinning out the "dead" channels, and also doing a ton of work that results in the 4th query running for several minutes without completing:

Running this query (which finds the counts for which no "live" channels exist):
```
materialize=> select count(distinct channel) from mz_message_counts c where not exists (select * from mz_dataflow_channels ch where ch.id = c.channel );
 count 
-------
    39
(1 row)
```

So ... it doesn't seem to be doing the thinning right?

And also, something is growing without bounds, and quickly:
```
materialize=> select now(), count(*) from mz_message_counts;
              now              | count 
-------------------------------+-------
 2021-03-02 21:41:57.537489+00 | 74816
(1 row)

materialize=> select now(), count(*) from mz_message_counts;
              now              | count  
-------------------------------+--------
 2021-03-02 21:42:04.034557+00 | 111104
(1 row)

materialize=> select now(), count(*) from mz_message_counts;
              now              | count  
-------------------------------+--------
 2021-03-02 21:42:07.930641+00 | 598528
(1 row)

materialize=> select now(), count(*) from mz_message_counts;
              now              | count  
-------------------------------+--------
 2021-03-02 21:42:10.531021+00 | 888832
(1 row)
```

**Comment by philip-stoev** [comment]
@antifuchs I will probably be one of the consumers for this information, so I have joined the review and will give this a try tomorrow.

My personal need is to be able to run a query, and then get a summary of the effort it took to evaluate the query so that I can compare e.g. two branches and decide if a particular optimization is beneficial or a net negative for various queries.

**Comment by antifuchs** [comment]
That sounds great, thank you @philip-stoev! I'll hopefully have something ready for you that won't crash out the materialized (overrunning counters) soon!

**Comment by antifuchs** [comment]
With the most recent change, the flow thins out the list of channels, but! It looks like the counts are getting inflated _incredibly_ fast - so fast that the materialized server crashes with a panic after 3 minutes (2:40 with `--debug-introspection`) because it's trying to add or multiply (it varies) i64's and is overflowing:

```
2021-03-03T20:28:38.588939Z  INFO dataflow::source::kafka: Activating Kafka queue for testdrive-input_proto-4106335075 [0] (source u3/26) on worker 4    
2021-03-03T20:31:19.756979Z ERROR panic: attempt to add with overflow
thread: timely:work-1
location: /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/core/src/ops/arith.rs:738:1
version: 0.7.1-dev (540b66f4dcad16c72a858ff07b6c64a5897e31b4)
backtrace:
   0: materialized::handle_panic
             at src/materialized/src/bin/materialized/main.rs:618:21
   1: core::ops::function::Fn::call
             at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/core/src/ops/function.rs:70:5
   2: std::panicking::rust_panic_with_hook
             at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/std/src/panicking.rs:597:17
   3: std::panicking::begin_panic_handler::{{closure}}
             at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/std/src/panicking.rs:497:13
   4: std::sys_common::backtrace::__rust_end_short_backtrace
             at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/std/src/sys_common/backtrace.rs:141:18
   5: rust_begin_unwind
             at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/std/src/panicking.rs:495:5
   6: core::panicking::panic_fmt
             at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/core/src/panicking.rs:92:14
   7: core::panicking::panic
             at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/core/src/panicking.rs:50:5
   8: <isize as core::ops::arith::AddAssign>::add_assign
             at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/core/src/ops/arith.rs:731:51
   9: <isize as core::ops::arith::AddAssign<&isize>>::add_assign
             at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/core/src/internal_macros.rs:73:17
  10: <differential_dataflow::difference::pair::DiffPair<R1,R2> as core::ops::arith::AddAssign<&differential_dataflow::difference::pair::DiffPair<R1,R2>>>::add_assign
             at /home/asf/.cargo/git/checkouts/differential-dataflow-d065d23d797aa027/e229d54/src/difference.rs:172:4
  11: differential_dataflow::consolidation::consolidate_slice
             at /home/asf/.cargo/git/checkouts/differential-dataflow-d065d23d797aa027/e229d54/src/consolidation.rs:62:17
  12: differential_dataflow::consolidation::consolidate_from
             at /home/asf/.cargo/git/checkouts/differential-dataflow-d065d23d797aa027/e229d54/src/consolidation.rs:30:18
  13: differential_dataflow::consolidation::consolidate
             at /home/asf/.cargo/git/checkouts/differential-dataflow-d065d23d797aa027/e229d54/src/consolidation.rs:21:5
  14: <differential_dataflow::operators::reduce::history_replay::HistoryReplayer<V1,V2,T,R1,R2> as differential_dataflow::operators::reduce::PerKeyCompute<V1,V2,T,R1,R2>>::compute
             at /home/asf/.cargo/git/checkouts/differential-dataflow-d065d23d797aa027/e229d54/src/operators/reduce.rs:915:25
  15: <differential_dataflow::operators::arrange::arrangement::Arranged<G,T1> as differential_dataflow::operators::reduce::ReduceCore<G,K,V,R>>::reduce_core::{{closure}}::{{closure}}
             at /home/asf/.cargo/git/checkouts/differential-dataflow-d065d23d797aa027/e229d54/src/operators/reduce.rs:530:49
  16: <timely::dataflow::stream::Stream<G,D1> as timely::dataflow::operators::generic::operator::Operator<G,D1>>::unary_frontier::{{closure}}::{{closure}}
             at /home/asf/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/84b1116/timely/src/dataflow/operators/generic/operator.rs:352:17
  17: timely::dataflow::operators::generic::builder_rc::OperatorBuilder<G>::build::{{closure}}::{{closure}}
             at /home/asf/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/84b1116/timely/src/dataflow/operators/generic/builder_rc.rs:123:31
  18: timely::dataflow::operators::generic::builder_rc::OperatorBuilder<G>::build_reschedule::{{closure}}
             at /home/asf/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/84b1116/timely/src/dataflow/operators/generic/builder_rc.rs:163:26
  19: <timely::dataflow::operators::generic::builder_raw::OperatorCore<T,L> as timely::scheduling::Schedule>::schedule
             at /home/asf/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/84b1116/timely/src/dataflow/operators/generic/builder_raw.rs:203:9
  20: timely::progress::subgraph::PerOperatorState<T>::schedule
             at /home/asf/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/84b1116/timely/src/progress/subgraph.rs:656:30
  21: timely::progress::subgraph::Subgraph<TOuter,TInner>::activate_child
             at /home/asf/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/84b1116/timely/src/progress/subgraph.rs:338:26
  22: <timely::progress::subgraph::Subgraph<TOuter,TInner> as timely::scheduling::Schedule>::schedule
             at /home/asf/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/84b1116/timely/src/progress/subgraph.rs:304:17
  23: timely::worker::Wrapper::step::{{closure}}
             at /home/asf/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/84b1116/timely/src/worker.rs:727:57
  24: core::option::Option<T>::map
             at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/core/src/option.rs:453:29
  25: timely::worker::Wrapper::step
             at /home/asf/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/84b1116/timely/src/worker.rs:727:26
  26: timely::worker::Worker<A>::step_or_park
             at /home/asf/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/84b1116/timely/src/worker.rs:375:38
  27: dataflow::server::Worker<A>::run
             at src/dataflow/src/server.rs:430:13
  28: dataflow::server::serve::{{closure}}
             at src/dataflow/src/server.rs:206:13
  29: timely::execute::execute::{{closure}}
             at /home/asf/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/84b1116/timely/src/execute.rs:284:22
  30: timely_communication::initialize::initialize_from::{{closure}}
             at /home/asf/.cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/84b1116/communication/src/initialize.rs:285:33
  31: std::sys_common::backtrace::__rust_begin_short_backtrace
             at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/std/src/sys_common/backtrace.rs:125:18
  32: std::thread::Builder::spawn_unchecked::{{closure}}::{{closure}}
             at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/std/src/thread/mod.rs:474:17
  33: <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/std/src/panic.rs:322:9
  34: std::panicking::try::do_call
             at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/std/src/panicking.rs:381:40
  35: __rust_try
  36: std::panicking::try
             at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/std/src/panicking.rs:345:19
  37: std::panic::catch_unwind
             at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/std/src/panic.rs:396:14
  38: std::thread::Builder::spawn_unchecked::{{closure}}
             at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/std/src/thread/mod.rs:473:30
  39: core::ops::function::FnOnce::call_once{{vtable.shim}}
             at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/core/src/ops/function.rs:227:5
  40: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
             at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/alloc/src/boxed.rs:1307:9
      <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
             at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/alloc/src/boxed.rs:1307:9
      std::sys::unix::thread::Thread::new::thread_start
             at /rustc/e1884a8e3c3e813aada8254edfa120e85bf5ffca/library/std/src/sys/unix/thread.rs:71:17
  41: start_thread
  42: __GI___clone
    
materialized encountered an internal error and crashed.
```

I think there are two options here:
* there's a bug and we aren't sending 171529281025560 messages over a channel (taken from a minute before the process crashed) [likely]
* maybe we should record rates instead of absolute numbers instead. [maybe avoid for now]

**Comment by frankmcsherry** [comment]
The issue is here:
```rust
        let messages = thin_collection(messages, delay, |c| {
             c.semijoin(&channels.map(|(channel, _)| channel))
         })
```
What you probably haven't been told and shouldn't have guessed is that `semijoin` on multisets is just defined (by differential, not math) as a join. So it multiplies the multiplicities, each time around.

To be honest, I'm not clear why this isn't a problem for the other logging code. @ruchirK do you know why?

**Comment by antifuchs** [comment]
Hah. We spent some time hashing this one out: The (probably too-simple) rephrasing is that we need to semijoin by the channel and the worker ID, which means we have to split out the sent & received messages into different tables so this becomes palatable (we can then make a view to make them look nicer again). 

I'll finish this up tomorrow, but I think the main problems&blockers for this PR are out of the way!

**Comment by philip-stoev** [comment]
@antifuchs can we easily get the number of bytes exchanged in addition to the message count? Or that needs to be a separate ticket/pr ?

File: src/dataflow-types/src/logging.rs [hunk 2]
```diff
@@ -116,6 +117,13 @@ impl LogVariant {
                 .with_column("count", ScalarType::Int64.nullable(false))
                 .with_key(vec![0, 1, 2]),
 
+            LogVariant::Timely(TimelyLog::Messages) => RelationDesc::empty()
+                .with_column("channel", ScalarType::Int64.nullable(false))
+                .with_column("source_worker", ScalarType::Int64.nullable(false))
+                .with_column("target_worker", ScalarType::Int64.nullable(false))
+                .with_column("sent", ScalarType::Int64.nullable(false))
+                .with_key(vec![0]),
```
Comment (ruchirK) [inline src/dataflow-types/src/logging.rs:125]: I'm not sure if just `channel` is sufficient to be the key? it seems like `channel`s are themselves keyed on`(id, worker)` so I think the appropriate key here is `(channel, source_worker, target_worker)`

File: src/dataflow-types/src/logging.rs [hunk 3]
```diff
@@ -189,6 +197,10 @@ impl LogVariant {
                 vec![(0, 0), (1, 1)],
             )],
             LogVariant::Timely(TimelyLog::Parks) => vec![],
+            LogVariant::Timely(TimelyLog::Messages) => vec![(
+                LogVariant::Timely(TimelyLog::Channels),
```
Comment (ruchirK) [inline src/dataflow-types/src/logging.rs:201]: this foreign key constraint seems maybe incorrect? the way I'm reading the code this is saying "Each row in messages must have a col0 (channel id) equal to the channel id of {some, exactly one} (I forget how foreign keys work) row in channels"? and then it says that twice?

I think if this is trying to reference the primary key of channels then we want logically

```
(messages.channel, messages.source_worker) = (channels.id, channels.worker)
```
as the fk constraint? but again, not totally sure whether foreign keys mean "an element with this key exists in the other relation" or "exactly one element with this key exists in the other relation".

Comment (ruchirK) [reply src/dataflow-types/src/logging.rs:201]: I suspect that might be related to why https://github.com/MaterializeInc/materialize/issues/5985 happens

File: src/dataflow-types/src/logging.rs (hunk 2 above)
Comment (antifuchs) [reply src/dataflow-types/src/logging.rs:125]: That was it, oof. I think channel, source_worker is probably sufficient to uniquely identify a channel, but (channel, source_worker, target_worker) is probably more accurate. Will make it so.

**Comment by antifuchs** [comment]
With 8dfa874, we now see the number of messages over a channel in /memory (looks like [this](https://materialize.com/memory-visualization/#eJydlm1vnDgQgP+KxX3JnmhkbBZwTqm0jTZSpKtySqL2Q6lWBLxbVC+sgL2+RPnvNzC2FxqgukSKhmGeGXvGM16enKxsnAsny3dVcvhCnuKCdH9xUR8f8V3spOpYN7LasKVLvNhpqU3o879ABCFFobUlihCF6ETkofBRBCiiTgh0FwwFIgIR0SEhpSgYCh9FgAIRDxEPEY+jQNLr9hIyRBgibIkC3bmHAv04huZo81u/51NZNkEQkjdviSkAiE6FQpBPKnmU6jJ2PLKvd3XsfO7yCKJxB1NArS6n/EPjoKu61Ko4OUCWQxeBjC59GA5VY/WnI4ANmWCGCTQTzTCYPBz0NCN0CQSbYZhm/KkyRZ4mdPfozQndaJgPdNOUv9DHRNnQYXJB66B70ThEEw6tpSO8qT20FiSmFoUe/x1BA02EehqGqt4lm96D7m8zKnSgUqNOtevJYZKgOgc9fuwXFfsbpvLkz4Yr6DqagfXG1WBqfa5Lokecm/lsR93MuZ50G+GmyOT3s5pTSl3y6apUx31xRhcu0Y/e4vOCnOXZBQG3BaykJ98GWFfV/4kRmRj9e+V9ctD2kFo7H7Vza+8dw6qGlZRMm7wsDLi0YDgP2qz6F8/tV8jLEMIQUe/orsoiTRpZwL/mIs9y/tjeI9/ae0d4rZKmxwSWiWbXsnUUo3UUto79y2dVVUmxk+9+3Eu1NSRzCSOVTMsqq8HJrZsfSl5uc6hT5qalKqvLCp7aF6jFzh/bLZw1DQKzRi/fO5kdU7lK0+P+CO+U2bHwXcJft44QZp1gNFdbMxG9yNUwpl6Di/JlO4DZgmweZBb050HfgsE8GFgwGuR5U/yjklQazKbijR09vLX2XgYP5X1TyWRvILt7j09D3EL+TDOC1XLhLGdGbXBLw86vE6Vy2ypgtRybi8dsEv07eYRbWm5uqMBquP4d/XJ/3LMcn4vHbf36d/agMcHikuXrxmK7NdGjsT7gNhufvlj9jayqsqo16sO9TV+3CUrNIrx/B+xObe1z+LquiVTJoZYZ7qn9LXp2XOffXH5zLp6cItlL+FzfQ+2qPFH5T3l+OD6qPD1Py2PRbMqDrMBWg0sKDdq08NXdevWwJh9u1h/hW77nGjvnsYPu3eMwBHznr+7J/frv9dUD+B12Gzi1RJW7Exs7Z38uyPXd7fs28M8BAGoG6laV33RIKGPsOM/P/wGZErbD)), and I think this change should make this PR ready to review - what do you all think?

**Comment by frankmcsherry** [comment]
I agree.

**Review by philip-stoev — APPROVED** [review]
Thank you. I am able to extract the counters and will try to make use of them for testing.

Can I open a new ticket for adding counters in bytes and not just in terms of messages? It seems that this information is also available and should be wired through. Thanks.

File: src/coord/src/catalog/builtin.rs [hunk 4]
```diff
@@ -1325,6 +1334,43 @@ lazy_static! {
 
         assert!(func_global_id_counter < 3000, "exhausted func global IDs");
 
+        // Ensure the IDs we assign are all unique:
```
Comment (ruchirK) [inline src/coord/src/catalog/builtin.rs:1337]: nit: if you could move this set of changes into its own commit I'd love that just because its logically distinct from everything else iuc. but also no pressure!

File: src/dataflow/src/logging/timely.rs [hunk 5]
```diff
@@ -374,6 +415,52 @@ pub fn construct<A: Allocate>(
                 }
             });
 
+        use differential_dataflow::operators::Count;
+        let messages_sent = thin_collection(messages_sent.as_collection(), delay, |c| {
```
Comment (ruchirK) [inline src/dataflow/src/logging/timely.rs:419]: I think all of this logic could use a few more comments. the two questions that jump out at me are:

1. why are sent and received two different collections? and
2. why are we thinning the collection?

File: src/dataflow/src/logging/timely.rs [hunk 6]
```diff
@@ -180,14 +186,14 @@ pub fn construct<A: Allocate>(
                                             for event in events {
```
Comment (ruchirK) [inline src/dataflow/src/logging/timely.rs:186]: this is entirely a random thought but I remember having a conversation about something adjacent to stuff like this where some changes made it so that we didn't have to keep around hashtables of the stream to replay it's negations back. looked everywhere on gh and can't find it so maybe I am misremembering

File: src/dataflow/src/logging/timely.rs [hunk 7]
```diff
@@ -294,7 +322,10 @@ pub fn construct<A: Allocate>(
 
         // Accumulate the durations of each operator.
         // The first `map` exists so that we can semijoin effectively (it requires a key-val pair).
-        let mut elapsed = duration
+        let mut elapsed: Collection<
```
Comment (ruchirK) [inline src/dataflow/src/logging/timely.rs:325]: is this change required or was this just for debugging purposes?

File: src/dataflow/src/logging/timely.rs [hunk 8]
```diff
@@ -345,8 +376,18 @@ pub fn construct<A: Allocate>(
             }
         });
 
-        let channels = channels.as_collection();
         let addresses = addresses.as_collection();
+        let channels = channels.as_collection().map(
+            |(id, worker, source_node, source_port, target_node, target_port)| {
+                (
+                    (id, worker),
```
Comment (ruchirK) [inline src/dataflow/src/logging/timely.rs:383]: I think you could just `session.give` the data with this structure to save on a map operator? not a huge deal

File: src/dataflow/src/logging/timely.rs (hunk 5 above)
Comment (ruchirK) [reply src/dataflow/src/logging/timely.rs:419]: ok my read on the answers for both is

1. it was simpler this way
2. to retain message counts for only the set of channels we are currently tracking?

I think thats all totally fine. it does seem to me like if we kept a is_sent variable around in the messages collection we could do the thinning with only one feedback loop / semijoin instead of two but idt thats worth optimizing for at this stage

**Review by ruchirK — APPROVED** [review]
these changes look great to me! left a few minor nits

File: src/dataflow/src/logging/timely.rs (hunk 5 above)
Comment (frankmcsherry) [reply src/dataflow/src/logging/timely.rs:419]: I think the answer is much more complicated, actually. The `thin` operation is only correct if the thing driving the thinning is correctly serialized with updates being thinned. I can't actually tell if this is happening here, but if the semijoin key is `(channel_id, worker)` then it is important that this record emerge before the thing it is semijoined with. Sent and received messages get logged at different workers (the sender and receiver, resp) and it is important to use their worker IDs.

@antifuchs does this sound right? Certainly worth a detailed comment if so, to prevent any future incorrect "simplification" of this code.

File: src/coord/src/catalog/builtin.rs (hunk 4 above)
Comment (antifuchs) [reply src/coord/src/catalog/builtin.rs:1337]: Yep, I'll rebase/squash this test and the extra table into two separate commits.

File: src/dataflow/src/logging/timely.rs (hunk 7 above)
Comment (antifuchs) [reply src/dataflow/src/logging/timely.rs:325]: That's debugging cruft, oops. I'll remove it.

File: src/dataflow/src/logging/timely.rs (hunk 8 above)
Comment (antifuchs) [reply src/dataflow/src/logging/timely.rs:383]: Huh, I think you're right. Will try!

File: src/dataflow/src/logging/timely.rs (hunk 5 above)
Comment (antifuchs) [reply src/dataflow/src/logging/timely.rs:419]: Yep, @frankmcsherry - that's correct. The join keys are different for data sent/received on channels, so we have to keep them separate, only later can we twist them together again.

File: src/coord/src/catalog/builtin.rs (hunk 4 above)
Comment (antifuchs) [reply src/coord/src/catalog/builtin.rs:1337]: done - it's three separate commits now (the test, the table, and the web UI), so I'll merge it as a non-squash.
